clash-prelude/src/Clash/XException.hs

```diff
@@ -99,6 +101,114 @@ infixr 0 `defaultSeqX`
+--
+-- > {-# LANGUAGE ViewPatterns, BangPatterns #-}
+-- > f (xToErrorCtx "a is X" -> !a) (xToErrorCtx "b is X" -> !b) = ...
+--
+-- __NB:__ Fully synthesizable, so doesn't have to be removed before synthesis
+--
+-- === __Example__
+--
+-- >>> :set -XViewPatterns -XDataKinds
+-- >>> import Clash.Sized.BitVector
+-- >>> import GHC.Stack
+-- >>> :{
+-- h, h' :: Bit -> BitVector 8 -> BitVector 8
+-- h (xToErrorCtx "a is X" -> a) (xToErrorCtx "b is X" -> b) = slice d7 d0 (pack a ++# b)
+-- h' a b = slice d7 d0 (pack a ++# b)
+-- :}
+--
+-- >>> h' (errorX "QQ") 3
+-- 0000_0011
+-- >>> h (errorX "QQ") 3
+-- *** Exception: a is X
+-- X: QQ
+-- CallStack (from HasCallStack):
+--   errorX, called at ...
+-- <BLANKLINE>
+xToErrorCtx :: String -> a -> a
+xToErrorCtx ctx a = unsafeDupablePerformIO
+  (catch (evaluate a >> return a)
+         (\(XException msg) ->
+           throw (ErrorCall (unlines [ctx,msg]))))
+{-# NOINLINE xToErrorCtx #-}
+
+-- | Convert 'XException' to 'ErrorCall'
+--
+-- This is useful when tracking the source of 'XException' that gets eaten up by
+-- 'Clash.Classes.BitPack.pack' inside of your circuit; since
+-- 'Clash.Classes.BitPack.pack' translates 'XException' into undefined bits.
+--
+-- So for example if you have some large function f:
+--
+-- > f a b = ... pack a ... pack b ...
+--
+-- Where it is basically an error if either /a/ or /b/ ever throws an 'XException',
+-- and so you want that to be reported the moment /a/ or /b/ is used, instead of
+-- it being thrown when evaluating the result of /f/, then do:
+--
+-- > {-# LANGUAGE ViewPatterns #-}
+-- > f (xToError -> a) (xToError -> b) = ...
+--
+-- Unlike 'xToErrorCtx', where we have an extra String argument to distinguish
+-- one call to 'xoToError' to the other, 'xToError' will use the 'GHC.CallStack'
+-- mechanism to aid the user in distinquishing different call to 'xToError'.
+-- We can also use BangPatterns to report the potential 'XException' being
+-- thrown by /a/ or /b/ even earlier, i.e. when /f/ is applied:
+--
+-- > {-# LANGUAGE ViewPatterns, BangPatterns #-}
+-- > f (xToError -> !a) (xToError -> !b) = ...
+--
+-- __NB:__ Fully synthesizable, so doesn't have to be removed before synthesis
+--
+-- === __Example__
+--
+-- >>> :set -XViewPatterns -XDataKinds
+-- >>> import Clash.Sized.BitVector
+-- >>> import GHC.Stack
+-- >>> :{
+-- f, g, h, h' :: HasCallStack => Bit -> BitVector 8 -> BitVector 8
+-- f = g
+-- g = h
+-- h (xToError -> a) (xToError -> b) = slice d7 d0 (pack a ++# b)
+-- h' a b = slice d7 d0 (pack a ++# b)
+-- :}
+--
+-- >>> h' (errorX "QQ") 3
+-- 0000_0011
+-- >>> f (errorX "QQ") 3
+-- *** Exception: CallStack (from HasCallStack):
+--   h, called at ...
+--   g, called at ...
+--   f, called at ...
+-- X: QQ
+-- CallStack (from HasCallStack):
+--   errorX, called at ...
+-- <BLANKLINE>
+xToError :: HasCallStack => a -> a
+xToError = xToErrorCtx (prettyCallStack (popCallStack callStack))
```

Comment:
Why do you `popCallStack`?
Doesn't that remove the removes the call that will tell me whether it was the `a` or the `b` argument to `h` that contained the XException?